triage/summarize/cluster.go

```diff
@@ -0,0 +1,296 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Contains functions that cluster failures.
+*/
+
+package summarize
+
+import "time"
+
+/*
+clusterLocal clusters together the failures for each test. Failures come in grouped
+by the test they belong to. These groups are subdivided into clusters.
+
+Takes:
+    {
+		testName1: [failure1, failure2, failure3, failure4, ...],
+		testName2: [failure5, failure6, failure7, failure8, failure9, ...],
+		...
+	}
+
+Returns:
+	{
+		testName1: {
+			clusterText1: [failure1, failure4, ...],
+			clusterText2: [failure3, ...],
+			clusterText3: [failure2, ...],
+			...
+		},
+		testName2: {
+			clusterText4: [failure5, failure7, failure8, ...],
+			clusterText5: [failure6, failure9, ...],
+			...
+		},
+		...
+	}
+*/
+func clusterLocal(failuresByTest failuresGroup) nestedFailuresGroups {
+	const memoPath string = "memo_cluster_local.json"
+	const memoMessage string = "clustering inside each test"
+
+	clustered := make(nestedFailuresGroups)
+
+	// Try to retrieve memoized results first to avoid another computation
+	if getMemoizedResults(memoPath, memoMessage, &clustered) {
+		return clustered
+	}
+
+	numFailures := 0
+	start := time.Now()
+	logInfo("Clustering failures for %d unique tests...", len(failuresByTest))
+
+	// Look at tests with the most failures first.
+	for n, pair := range failuresByTest.sortByNumberOfFailures() {
+		numFailures += len(pair.Failures)
+		logInfo("%4d/%4d tests, %5d failures, %s", n+1, len(failuresByTest), len(pair.Failures), pair.Key)
+		clustered[pair.Key] = clusterTest(pair.Failures)
+	}
+
+	elapsed := time.Since(start)
+	logInfo("Finished locally clustering %d unique tests (%d failures) in %s", len(clustered), numFailures, elapsed.String())
+
+	// Memoize the results
+	memoizeResults(memoPath, memoMessage, clustered)
+	return clustered
+}
+
+// failure represents a specific instance of a test failure.
+type failure struct {
+	Started     int    `json:"started"`
+	Build       string `json:"build"`
+	Name        string `json:"name"`
+	FailureText string `json:"failure_text"`
+}
```

Comment:
I wonder if there's a better place to put this type? Seems randomly sandwiched in between two funcs

Reply:
I tried to follow Go standards and place it before its first use. It technically gets used in files.go first, but that's not its main use so I put it here instead.